### ISSUE TITLE
Apply selector-pseudo-class-*-list rules to at-page

### ISSUE THREAD

**Issue by alexander-akait** [stylelint/stylelint#1871]
- [x] `selector-pseudo-class-no-unknown`
- [x] `selector-pseudo-class-disallowed-list`
- [ ] `selector-pseudo-class-allowed-list`

> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

https://developer.mozilla.org/en-US/docs/Web/CSS/:first
https://developer.mozilla.org/en-US/docs/Web/CSS/:right
https://developer.mozilla.org/en-US/docs/Web/CSS/:left
https://developer.mozilla.org/en-US/docs/Web/CSS/:blank

cf https://github.com/stylelint/stylelint/blob/dfd1ffcabb2036d9e2e25aed512c193d3d868ae5/lib/reference/selectors.js#L197

> Which rule, if any, is this issue related to?

`selector-pseudo-class-*`

> What CSS is needed to reproduce this issue?

e.g.

``` css
@page :first {
  margin-left: 50%;
  margin-top: 50%;
}
```

> What stylelint configuration is needed to reproduce this issue?

No matter.

> Which version of stylelint are you using?

No matter.

> How are you running stylelint: CLI, PostCSS plugin, Node API?

No matter.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No matter.

> What did you expect to happen?

Check `pseudo-classes` in this rules

> What actually happened (e.g. what warnings or errors you are getting)?

Now we have not check.

**Comment by jeddy3** [comment]
Have you run into a real world need for this?

**Comment by evilebottnawi** [comment]
@jeddy3 yep, for print styles, also this include in spec 2.1, https://developer.mozilla.org/ru/docs/Web/CSS/@page

**Comment by jeddy3** [comment]
> also this include in spec 2.1,

Oh wow, I had in mind that this was something new. In fact, it's spec 2.1 and [with excellent browser support](https://developer.mozilla.org/en/docs/Web/CSS/@page#Browser_compatibility).

LGTM for `future-major`.

**Comment by davidtheclark** [comment]
This should be finished for v8 or closed.

**Comment by jeddy3** [comment]
@evilebottnawi Am I right in thinking the following rules also need  at-page support?:

- `selector-pseudo-class-blacklist`
- `selector-pseudo-class-case`
- `selector-pseudo-class-parentheses-space-inside`
- `selector-pseudo-class-whitelist`

**Comment by evilebottnawi** [comment]
@jeddy3 yep, I'll try to do this on weekend

**Comment by jeddy3** [comment]
> yep, I'll try to do this on weekend

Awesome. I'm going to try to make some time tomorrow to tackle a couple of the outstanding issues too. Hopefully, we'll be able to push out `8.0.0` soon :)

**Comment by jeddy3** [comment]
With the semantic versioning change in `8.0.0`, I think this can be done in an `8.x`.

@evilebottnawi Feel free to get it in `8.0.0`, but I don't think it's a prerequisite for the `8.0.0` release.

**Comment by jonathantneal** [comment]
We very recently adopted stylelint and are now rolling it out across multiple sites, including multiple documentation sites. Thanks for a great project that’s always been super helpful!

I totally understand that print CSS is less common. However, as all of our documentation sites are print-friendly, this issue is tripping us up.

The work-around is to either 1. add the following rule to each documentation project, or 2. add it into a package to then deploy into every project.

```js
{
  rules: {
    'selector-pseudo-class-no-unknown': [
      true,
      {
        ignorePseudoClasses: ['first', 'last']
      }
    ]
  }
}
```

Is there a better workaround? My concern is that I’m not a fan of allowing `:first` or `:last` on just _any_ selector.

**Comment by jeddy3** [comment]
It's fantastic that you're rolling out stylelint and you're finding it helpful.

(As an aside, your recent work on the experimental PostCSS tokenizer is very exciting and could be immensely beneficial for lots of projects, including stylelint.) 

> Is there a better workaround? My concern is that I’m not a fan of allowing :first or :last on just any selector.

I can't reproduce the `selector-pseudo-class-no-unknown` issue [using the demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIACAHAhgcxgAngMwEsAnAZwBdtgBfAHQDtVMd4ALCANxiMtrvrVYcuPfvmLkRdEABoQJAJ50yaAB4JwJEjPAQ6hDOuD1s2GiCIBXADYwSZ+JWMnTcmDbBkIRALQoSMCygIbzArNE1vOmCLOgBrKIB3PgRsMksYJ14qECogA):

```css
@page :first {}
@page :hover {}

a:hover {}
a:first {}
```

Gives me:

```shell
2:7 error Unexpected unknown pseudo-class selector ":hover" (selector-pseudo-class-no-unknown)
5:2 error Unexpected unknown pseudo-class selector ":first" (selector-pseudo-class-no-unknown)
```

Where `:hover` is a violation for the at-page but not the rule, and `:first` for the rule but not the at-page.

Support for at-page pseudo-classes was added to the `selector-pseudo-class-no-unknown` rule back in [`7.10.0`](https://stylelint.io/changelog#7110). We still need to add at-page support for other `selector-pseudo-class-*` rules, like `selector-pseudo-class-blacklist` and `selector-pseudo-class-case`, which is why this issue remains open.

Are you seeing different results locally for the `selector-pseudo-class-no-unknown` rule?

> I totally understand that print CSS is less common.

One of the [goals of stylelint](https://stylelint.io/about/vision) is to support all of standard CSS, and print CSS is very much part of that. I don't think we can consider it finished until it does :)

**Comment by jeddy3** [comment]
Removed stylistic rules from task list (see https://github.com/stylelint/stylelint/issues/5674).

Two remain:

- `selector-pseudo-class-disallowed-list`
- `selector-pseudo-class-allowed-list`

**Comment by Mouvedia** [comment]
I've labeled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

**Comment by Mouvedia** [comment]
@firefoxic 

- [`selector-pseudo-class-case`](https://github.com/stylelint-stylistic/stylelint-stylistic/tree/main/lib/rules/selector-pseudo-class-case)
- [`selector-pseudo-class-parentheses-space-inside`](https://github.com/stylelint-stylistic/stylelint-stylistic/tree/main/lib/rules/selector-pseudo-class-parentheses-space-inside)
